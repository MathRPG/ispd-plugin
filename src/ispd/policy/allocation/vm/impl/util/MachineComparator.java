package ispd.policy.allocation.vm.impl.util;

import ispd.motor.filas.servidores.CS_Processamento;
import java.io.Serializable;
import java.util.Comparator;

public abstract class MachineComparator <T extends CS_Processamento>
    implements Comparator<T>, Serializable {

    protected abstract int calculateMachineValue (T m);

    @Override
    public int compare (final T m1, final T m2) {
        return this.calculateMachineValue(m1) - this.calculateMachineValue(m2);
    }
}
