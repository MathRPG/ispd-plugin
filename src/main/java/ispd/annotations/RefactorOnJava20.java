package ispd.annotations;

import java.lang.annotation.*;

/**
 * Indicates that a method or constructor should be refactored once the project is upgraded to use
 * the Java 20 JDK.
 */
@Retention(RetentionPolicy.SOURCE)
@Target({ ElementType.METHOD, ElementType.CONSTRUCTOR })
public @interface RefactorOnJava20 {

    /**
     * @return The reason for the potential refactoring, such as {@code "Use Pattern Matching."}
     */
    String reason() default "";
}
