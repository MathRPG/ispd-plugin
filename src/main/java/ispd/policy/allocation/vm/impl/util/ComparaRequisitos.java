package ispd.policy.allocation.vm.impl.util;

import ispd.motor.queues.centers.impl.*;

public class ComparaRequisitos extends MachineComparator<VirtualMachine> {

    private static final int MULTIPLIER = 100_000;

    protected int calculateMachineValue (final VirtualMachine m) {
        return MULTIPLIER * (
            m.getProcessadoresDisponiveis()
            + (int) m.getMemoriaDisponivel()
            + (int) m.getDiscoDisponivel()
        );
    }
}
