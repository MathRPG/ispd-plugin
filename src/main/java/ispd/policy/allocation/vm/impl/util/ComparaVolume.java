package ispd.policy.allocation.vm.impl.util;

import ispd.motor.queues.centers.impl.*;

public class ComparaVolume extends MachineComparator<CloudMachine> {

    private static final int VOLUME = 1_000_000;

    protected int calculateMachineValue (final CloudMachine m) {
        return VOLUME * (
            m.getProcessadoresDisponiveis()
            * (int) m.getMemoriaDisponivel()
            * (int) m.getDiscoDisponivel()
        );
    }
}
