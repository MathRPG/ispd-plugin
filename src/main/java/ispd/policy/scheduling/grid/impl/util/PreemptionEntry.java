package ispd.policy.scheduling.grid.impl.util;

import ispd.motor.queues.task.*;

public record PreemptionEntry(
    String preemptedTaskUser, int preemptedTaskId, String scheduledTaskUser, int scheduledTaskId
) {

    public PreemptionEntry (final GridTask preempted, final GridTask scheduled) {
        this(preempted.getProprietario(), preempted.getIdentificador(), scheduled.getProprietario(),
             scheduled.getIdentificador()
        );
    }

    public boolean willPreemptTask (final GridTask task) {
        return this.preemptedTaskId == task.getIdentificador()
               && this.preemptedTaskUser.equals(task.getProprietario());
    }

    public boolean willScheduleTask (final GridTask task) {
        return this.scheduledTaskId == task.getIdentificador()
               && this.scheduledTaskUser.equals(task.getProprietario());
    }
}
