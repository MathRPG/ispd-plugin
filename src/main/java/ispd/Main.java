package ispd;

import ispd.application.*;
import ispd.application.terminal.*;
import java.util.*;

public enum Main {
    ;

    private static final Locale EN_US_LOCALE = new Locale("en", "US");

    public static void main (final String[] args) {
        setDefaultLocale();

        final var app = (args.length == 0)
                        ? new GuiApplication()
                        : new TerminalApplication(args);

        app.run();
    }

    private static void setDefaultLocale () {
        Locale.setDefault(Main.EN_US_LOCALE);
    }
}
