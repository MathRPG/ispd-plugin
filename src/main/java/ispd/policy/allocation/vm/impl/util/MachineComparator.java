package ispd.policy.allocation.vm.impl.util;

import ispd.motor.queues.centers.*;
import java.io.*;
import java.util.*;

public abstract class MachineComparator <T extends Processing>
    implements Comparator<T>, Serializable {

    protected abstract int calculateMachineValue (T m);

    @Override
    public int compare (final T m1, final T m2) {
        return this.calculateMachineValue(m1) - this.calculateMachineValue(m2);
    }
}
