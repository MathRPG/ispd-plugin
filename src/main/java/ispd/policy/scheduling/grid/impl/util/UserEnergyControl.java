package ispd.policy.scheduling.grid.impl.util;

import ispd.motor.queues.centers.*;
import java.util.*;

public class UserEnergyControl extends UserProcessingControl {

    private final double energyEfficiencyRatioAgainstSystem;

    private final double energyConsumptionLimit;

    private final double ownedMachinesEnergyConsumption;

    private double currentEnergyConsumption = 0.0;

    public UserEnergyControl (
        final String userId,
        final Collection<? extends Processing> systemMachines,
        final double energyConsPercentage
    ) {
        super(userId, systemMachines);

        this.ownedMachinesEnergyConsumption = this
            .ownedNonMasterMachinesIn(systemMachines)
            .mapToDouble(Processing::getConsumoEnergia)
            .sum();

        this.energyConsumptionLimit =
            this.calculateEnergyConsumptionLimit(energyConsPercentage);

        this.energyEfficiencyRatioAgainstSystem =
            this.calculateEnergyEfficiencyRatioAgainst(systemMachines);
    }

    @Override
    public boolean isEligibleForTask () {
        return super.isEligibleForTask()
               && !this.hasExceededEnergyLimit();
    }

    @Override
    public void stopTaskFrom (final Processing machine) {
        super.stopTaskFrom(machine);
        this.decreaseEnergyConsumption(machine.getConsumoEnergia());
    }

    @Override
    public void startTaskFrom (final Processing machine) {
        super.startTaskFrom(machine);
        this.increaseEnergyConsumption(machine.getConsumoEnergia());
    }

    private double calculateEnergyConsumptionLimit (final double energyConsPercentage) {
        return this.ownedMachinesEnergyConsumption * energyConsPercentage / 100;
    }

    private double calculateEnergyEfficiencyRatioAgainst (
        final Collection<? extends Processing> machines
    ) {
        final var sysComputationPower = machines.stream()
            .mapToDouble(Processing::getPoderComputacional)
            .sum();

        final var sysEnergyConsumption = machines.stream()
            .mapToDouble(Processing::getConsumoEnergia)
            .sum();

        final var sysEnergyEff = sysComputationPower / sysEnergyConsumption;
        return sysEnergyEff / this.energyEfficiency();
    }

    private double energyEfficiency () {
        return this.getOwnedMachinesProcessingPower() / this.ownedMachinesEnergyConsumption;
    }

    private void increaseEnergyConsumption (final double amount) {
        this.currentEnergyConsumption += amount;
    }

    private void decreaseEnergyConsumption (final double amount) {
        this.currentEnergyConsumption -= amount;
    }

    private boolean hasExceededEnergyLimit () {
        return this.currentEnergyConsumption >= this.energyConsumptionLimit;
    }

    public double currentConsumptionWeightedByEfficiency () {
        return this.currentEnergyConsumption * this.energyEfficiencyRatioAgainstSystem;
    }

    public boolean hasLessEnergyConsumptionThan (final UserEnergyControl other) {
        return this.energyConsumptionLimit <= other.energyConsumptionLimit;
    }

    public boolean canUseMachineWithoutExceedingEnergyLimit (final Processing machine) {
        return this.currentEnergyConsumption + machine.getConsumoEnergia()
               <= this.energyConsumptionLimit;
    }

    public double getOwnedMachinesEnergyConsumption () {
        return this.ownedMachinesEnergyConsumption;
    }
}
