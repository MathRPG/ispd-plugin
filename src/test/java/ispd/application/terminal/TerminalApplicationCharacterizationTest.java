package ispd.application.terminal;

import static ispd.application.terminal.HasMessageIn.*;
import static org.approvaltests.Approvals.*;
import static org.hamcrest.MatcherAssert.*;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.*;

import ispd.annotations.*;
import java.io.*;
import java.net.*;
import java.nio.charset.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import org.apache.commons.cli.*;
import org.hamcrest.core.*;
import org.jetbrains.annotations.*;
import org.junit.jupiter.api.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

class TerminalApplicationCharacterizationTest {

    public static final Path MODEL_FOLDER_PATH = Path.of("src", "test", "resources", "models");

    private static final String[] NO_OPTIONS = {};

    private static final Pattern SPACE_MATCHER = Pattern.compile(" ");

    private final PrintStream standardOut = System.out;

    private final ByteArrayOutputStream outStream = new ByteArrayOutputStream();

    @RefactorOnJava20(reason = "Use Pattern Matching.")
    private static String @Nullable [] convertToOptionArray (final CharSequence spacedOptions) {
        if (spacedOptions == null) {
            return null;
        }
        if (spacedOptions.isEmpty()) {
            return NO_OPTIONS;
        }
        return SPACE_MATCHER.split(spacedOptions);
    }

    private static TerminalApplication initTerminalApplication (final CharSequence spacedOptions) {
        return new TerminalApplication(convertToOptionArray(spacedOptions));
    }

    private static void runTerminalApplication (final CharSequence spacedOptions) {
        initTerminalApplication(spacedOptions).run();
    }

    private static @NotNull String makePathToModel (final String modelName) {
        return Path.of("src", "test", "resources", "models", modelName).toString();
    }

    private @NotNull Map<String, Object> mapOfExceptionAndOut (final Exception exception) {
        return Map.of(
            "ex", exception, // TODO: Change
            "out", this.outStream
        );
    }

    private String systemOutContents () {
        return this.outStream.toString();
    }

    @BeforeEach
    void replaceSystemOut () {
        System.setOut(new PrintStream(this.outStream, true, StandardCharsets.UTF_8));
    }

    @AfterEach
    void reinstateStandardSystemOut () {
        System.setOut(this.standardOut);
    }

    @ParameterizedTest
    @NullAndEmptySource
    void givenEmptyOrNullArgs_thenThrowsOnInit (final String options) {
        final var exception = assertThrows(
            IllegalArgumentException.class,
            () -> initTerminalApplication(options)
        );

        assertThat(exception, hasMessageIn(this.systemOutContents()));

        verify(this.systemOutContents());
    }

    @Test
    void givenUnrecognizedOption_thenThrowsOnInit () {
        final var cause = assertThrows(
            RuntimeException.class,
            () -> initTerminalApplication("-z")
        ).getCause();

        assertThat(
            cause,
            both(hasMessageIn(this.systemOutContents()))
                .and(is(instanceOf(UnrecognizedOptionException.class)))
        );

        verify(this.systemOutContents());
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "-P",
            "--port",
            "-e",
            "-t",
            "-a",
        }
    )
    void givenOptionWithMissingArgument_thenThrowsOnInit (final String options) {
        final var cause = assertThrows(
            RuntimeException.class,
            () -> initTerminalApplication(options)
        ).getCause();

        assertThat(
            cause,
            both(hasMessageIn(this.systemOutContents()))
                .and(is(instanceOf(MissingArgumentException.class)))
        );
    }

    @Test
    void givenOptionWithMissingArgument_thenPrintsErrorToOut () {
        try {
            initTerminalApplication("-P");
        } catch (final RuntimeException ignored) {
            // ... throwing behavior already tested
        }

        verify(this.systemOutContents());
    }

    @Test
    void givenInvalidAddress_thenThrowsOnInit () {
        final var cause = assertThrows(
            IllegalArgumentException.class,
            () -> initTerminalApplication("-a NotAnAddress")
        ).getCause();

        assertThat(cause, this.hasMessageInSystemOut_andIsOfType(UnknownHostException.class));

        verify(this.systemOutContents());
    }

    private <T> CombinableMatcher<Throwable> hasMessageInSystemOut_andIsOfType (final Class<T> type) {
        return both(hasMessageIn(this.systemOutContents()))
            .and(is(instanceOf(type)));
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "-P NaN",
            "-e NaN",
            "-t NaN",
        }
    )
    void givenOptionWithInvalidNumberArgument_thenThrowsOnInit (final String options) {
        final var cause = assertThrows(
            RuntimeException.class,
            () -> initTerminalApplication(options)
        ).getCause();

        assertThat(
            cause,
            is(instanceOf(NumberFormatException.class))
        );

        verify(this.systemOutContents());
    }

    @Test
    void givenValidArgs_thenPrintsNothingToOutOnInit () {
        initTerminalApplication("-h");

        assertThat(
            "Should not print anything to out on valid initialization.",
            this.systemOutContents(),
            is(emptyString())
        );
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "-h",
            "--help",
            "-h -v",
            "-v -h",
            "-h doesNotExist.imsx",
            "doesNotExist.imsx -h",
        }
    )
    void givenHelpArg_whenRun_thenPrintsHelp (final String options) {
        runTerminalApplication(options);

        verify(this.outStream);
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "-v",
            "--version",
            "-v doesNotExist.imsx",
            "doesNotExist.imsx -v",
        }
    )
    void givenVersionArg_whenRun_thenPrintsVersionInfo (final String options) {
        runTerminalApplication(options);

        verify(this.outStream);
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "nonexistent.imsx",
            "wrongExtension",
            "wrongExtension.txt",
        }
    )
    void givenNonexistentOrWrongExtensionModel_thenPrintsErrorToOutOnRun (final String modelName) {
        runTerminalApplication(ModelFolder.NO_TYPE.pathTo(modelName));

        assertThat(
            "Error message printed to out should contain model name",
            this.systemOutContents(),
            containsString(modelName)
        );

        verify(
            this.outStream.toString().replace(modelName, "")
        );
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            "emptyFile.imsx",
            // Grid models
            "gridMalformedModel.imsx",
            "emptyGridModel.imsx",
            "emptyGridModelWithSingleUser.imsx",
            "gridModelWithSingleMachineIcon.imsx",
            "gridModelWithSingleTask.imsx",
            // Iaas models
            "emptyIaaSModel.imsx",
            "emptyIaaSModelWithSingleUser.imsx",
            "iaasModelWithSingleMachineIcon.imsx",
            "iaasModelWithSingleTask.imsx",
            // Paas models
            "emptyPaaSModel.imsx",
            "emptyPaaSModelWithSingleUser.imsx",
            "paasModelWithSingleMachineIcon.imsx",
            "paasModelWithSingleTask.imsx",
        }
    )
    void givenInvalidModel_whenRun_thenPrintsErrors (final String modelName) {
        runTerminalApplication(makePathToModel(modelName));

        verify(this.outStream, NAMES.withParameters(modelName));
    }

    @ParameterizedTest
    @ValueSource(
        strings = {
            // Grid Models
            "gridModelWithSingleMaster.imsx",
            "gridModelWithNoSlaves.imsx",
            // IaaS Models
            "iaasModelWithSingleMaster.imsx",
            "iaasModelWithInvalidAllocation.imsx",
            "iaasModelWithNoSlaves.imsx",
            // PaaS Models
            "paasModelWithSingleMaster.imsx",
            "paasModelWithInvalidAllocation.imsx",
            "paasModelWithNoSlaves.imsx",
        }
    )
    void givenModelWithInvalidPolicies_whenRun_thenThrowsException (final String modelName) {
        final var path = makePathToModel(modelName);

        final var exception = assertThrows(
            RuntimeException.class,
            () -> runTerminalApplication(path)
        );

        verify(
            this.mapOfExceptionAndOut(exception),
            NAMES.withParameters(modelName)
        );
    }
}
