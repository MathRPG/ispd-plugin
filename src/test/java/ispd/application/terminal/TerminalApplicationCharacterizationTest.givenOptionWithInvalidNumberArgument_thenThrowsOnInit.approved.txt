"NaN" is not a valid number
